accepted/future-releases/inline-classes/feature-specification.md

```diff
@@ -0,0 +1,948 @@
+  tiny.add("Hello!"); // Error.
+}
+```
+
+This is similar to the inline type in that it enforces the use of
+specific operations only (here we have just one: `leaves`), and it
+makes it an error to use instance methods of the representation (e.g.,
+`add`).
+
+Creation of wrapper objects consumes space and time. In the case where
+we wish to work on an entire data structure, we'd need to wrap each
+object as we navigate the data structure. For instance, we need to
+create a wrapper `TinyJson(element)` in order to invoke `leaves`
+recursively.
+
+In contrast, the inline class is zero-cost, in the sense that it does
+_not_ use a wrapper object, it enforces the desired discipline
+statically. In the inline class, the invocation of `TinyJson(element)`
+in the body of `leaves` can be eliminated entirely by inlining.
+
+Inline classes are static in nature, like extension members: an inline
+class declaration may declare some type parameters. The type
+parameters will be bound to types which are determined by the static
+type of the receiver. Similarly, members of an inline type are resolved
+statically, i.e., if `tiny.leaves` is an invocation of an inline getter
+`leaves`, then the declaration named `leaves` whose body is executed
+is determined at compile-time. There is no support for late binding of
+an inline member, and hence there is no notion of overriding. In return
+for this lack of expressive power, we get improved performance.
+
+
+## Syntax
+
+A rule for `<inlineClassDeclaration>` is added to the grammar, along
+with some rules for elements used in inline class declarations:
+
+```ebnf
+<inlineClassDeclaration> ::=
+  'inline' 'class' <typeIdentifier> <typeParameters>? <interfaces>?
+  '{'
+    (<metadata> <inlineMemberDeclaration>)*
+  '}'
+
+<inlineMemberDeclaration> ::= <classMemberDefinition>
+```
+
+*The token `inline` is not made a built-in identifier: the reserved
+word `class` that occurs right after `inline` serves to disambiguate
+the inline class declaration with a fixed lookahead.*
+
+A few errors can be detected immediately from the syntax:
+
+A compile-time error occurs unless the inline class declares exactly
+one instance variable. Let `v` be the name of said instance
+variable. A compile-time error occurs unless the declaration of `v`
+has the modifier `final` or the modifier `late`.
+
+The _name of the representation_ in an inline class declaration is the
+name `id` of the unique final instance variable that it declares, and
+the _type of the representation_ is the declared type of `id`.
+
+A compile-time error occurs if an inline class declaration declares an
+abstract member.
+
+*There are no special rules for static members in inline classes. They
+can be declared and called or torn off as usual, e.g.,
+`Inline.myStaticMethod(42)`.*
+
+
+## Primitives
+
+This document needs to refer to explicit inline method invocations, so we
+will add a special primitive, `invokeInlineMethod`, to denote invocations of
```

Comment:
I've commented on this in previous PRs so I'm going to try to keep this brief, but I still think it's a bad idea to define the static analysis in terms of desugaring to a notation that's invalid Dart.  The actual implementation of the static analysis (at least in the analyzer) will not use this desugaring, so although we might gain extra precision in the specification by specifying it in this way, we pay for that precision by complicating the job of the implementers, who will have to mentally translate every line of the text that follows to figure out what it means in terms of the code the user can actually write.  IMHO it would be better for the spec to be written in terms of the code the user can actually write, because it would reduce the risk of implementation mistakes.

Reply:
I believe the CFE team has given similar feedback in the past.